### PR TITLE
Update test_hashlib.py from Cpython v3.11.2 & refactor hashlib

### PR DESCRIPTION
**Pull request by Masorubka1**
1) split hashlib and _hashlib
2) add libraries: _md5, _sha1, _sha3, _sha256, _blake2
3) add tests for hashlib

### CODE REVIEW

**Comment by fanninpm** [comment]
Hmm… I think we need to split up [`stdlib/src/hashlib.rs`](https://github.com/RustPython/RustPython/blob/main/stdlib/src/hashlib.rs) into different modules and turn `hashlib` into a pure-Python module. What do you think, @youknowone?

**Comment by youknowone** [comment]
`hashlib.py` strongly supposes `_hashlib` is cpython-like openssl implementation. It will require efforts.
In my opinion, that's more like test-purpose compatibility  rather than actual compatiblity. But I will happily take patch if anyone want to work on it.

**Comment by Masorubka1** [comment]
@youknowone Could you review this pr?

File: stdlib/Cargo.toml [hunk 1]
```diff
@@ -55,6 +55,7 @@ sha-1 = "0.10.0"
 sha2 = "0.10.2"
 sha3 = "0.10.1"
 blake2 = "0.10.4"
+hmac = {version = "0.12.1"}
```
Comment (fanninpm) [inline stdlib/Cargo.toml:58]: Simplification:
```suggestion
hmac = "0.12.1"
```

File: stdlib/src/md5.rs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+pub(crate) use _md5::make_module;
+
+#[pymodule]
+mod _md5 {
+    use crate::hashlib::_hashlib::{HashArgs, HashWrapper, PyHasher};
+    use crate::vm::{PyObjectRef, PyPayload, PyResult, VirtualMachine};
+    use md5::Md5;
+
+    #[pyfunction(name = "md5")]
+    fn md5(args: HashArgs, vm: &VirtualMachine) -> PyResult<PyObjectRef> {
```
Comment (youknowone) [inline stdlib/src/md5.rs:10]:
```suggestion
    fn md5(args: HashArgs, vm: &VirtualMachine) -> PyResult {
```
PyResult = PyResult<PyObjectRef>

**Review by youknowone — CHANGES_REQUESTED** [review]

File: stdlib/src/md5.rs [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+        Ok(PyHasher::new("md5", HashWrapper::new::<Md5>(args.string)).into_pyobject(vm))
```
Comment (youknowone) [inline stdlib/src/md5.rs:11]:
```suggestion
        Ok(crate::hashlib::md5(args).into_pyobject(vm))
```
Is this possible?

File: stdlib/src/hashlib.rs [hunk 4]
```diff
@@ -229,90 +231,142 @@ mod hashlib {
...
+        };
+
+        let b_hash = match b.class().is(_py_str) {
+            true => b.try_to_value::<&str>(vm)?.as_bytes().to_vec(),
+            _ => b.try_to_value::<PyBytes>(vm)?.as_bytes().to_vec(),
+        };
+
+        println!("\na: {:?}", a_hash);
+        println!("b: {:?}", b_hash);
+        println!("equal: {}", (a_hash == b_hash));
```
Comment (youknowone) [inline stdlib/src/hashlib.rs:336]: debug code remained

File: stdlib/src/hashlib.rs [hunk 5]
```diff
@@ -229,90 +231,142 @@ mod hashlib {
...
+    #[pyfunction(name = "openssl_blake2s")]
     fn blake2s(args: BlakeHashArgs) -> PyHasher {
         PyHasher::new("blake2s", HashWrapper::new::<Blake2s256>(args.data))
     }
 
-    trait ThreadSafeDynDigest: DynClone + DynDigest + Sync + Send {}
+    #[pyfunction]
+    fn compare_digest(
+        a: PyObjectRef,
+        b: PyObjectRef,
```
Comment (youknowone) [inline stdlib/src/hashlib.rs:307]:
```suggestion
        a: ArgStrOrBytesLike,
        b: ArgStrOrBytesLike,
```
This type will be helpful

**Review by youknowone — CHANGES_REQUESTED** [review]
Wow, You made it 👍 
Looks mostly good. I left a few comments.

File: Lib/test/test_enum.py [hunk 6]
```diff
@@ -2889,9 +2889,7 @@ class Color(StrMixin, AllMixin, IntFlag):
         self.assertEqual(Color.ALL.value, 7)
         self.assertEqual(str(Color.BLUE), 'blue')
 
-    # TODO: RUSTPYTHON
-    @unittest.expectedFailure
-    @unittest.skipIf(sys.platform == "win32", "TODO: RUSTPYTHON, inconsistent test result on Windows due to threading")
+    @unittest.skipIf(sys.platform == "win32" or sys.platform.startswith("linux"), "TODO: RUSTPYTHON, inconsistent test result on Windows due to threading")
```
Comment (fanninpm) [inline Lib/test/test_enum.py:2892]: If we also observe this on macOS…

```suggestion
    @unittest.skip("TODO: RUSTPYTHON, inconsistent test result due to threading")
```

File: Lib/test/test_secrets.py [hunk 7]
```diff
@@ -44,6 +44,7 @@ def test_bad_types(self):
 
     def test_bool(self):
         # Test that compare_digest returns a bool.
+        print(secrets.compare_digest("abc", "abc"))
```
Comment (youknowone) [inline Lib/test/test_secrets.py:47]:
```suggestion
```
Unless the print came from CPython source code

File: vm/src/function/buffer.rs [hunk 8]
```diff
@@ -141,6 +141,20 @@ impl ArgStrOrBytesLike {
...
+            Self::Str(_) => "str",
+        }
+    }
+
+    pub fn get_enum_int(&self) -> u8 {
+        match self {
+            Self::Buf(_) => 1,
+            Self::Str(_) => 2,
+        }
+    }
```
Comment (youknowone) [inline vm/src/function/buffer.rs:157]: This is too much ad-hoc

File: stdlib/src/hashlib.rs [hunk 9]
```diff
@@ -211,108 +213,144 @@ mod hashlib {
...
+        b: ArgStrOrBytesLike,
+        vm: &VirtualMachine,
+    ) -> PyResult<PyObjectRef> {
+        if a.get_enum_int() != b.get_enum_int() {
+            return Err(vm.new_type_error(format!(
+                "different types in compare_digest {} and {} ",
+                a.get_enum(),
+                b.get_enum()
+            )));
+        }
```
Comment (youknowone) [inline stdlib/src/hashlib.rs:316]: The error message is different.
I am not going to ask to match the message matching, but I disagree about adding `get_enum` and `get_enum_int` to make incompatible error message.
```
Python 3.11.1 (main, Feb  4 2023, 11:11:18) [Clang 14.0.0 (clang-1400.0.29.202)] on darwin
>>> _hashlib.compare_digest('', bytearray(b''))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: a bytes-like object is required, not 'str'
```

```suggestion
        fn is_str(arg: &ArgStrOrBytesLike) -> {
           matches!(arg, ArgStrOrBytesLike::String(_))
        }
        if is_str(&a) != is_str(&b) {
            return Err(...);
        }
```

**Review by youknowone — CHANGES_REQUESTED** [review]

File: stdlib/src/hashlib.rs (hunk 9 above)
Comment (Masorubka1) [reply stdlib/src/hashlib.rs:316]: @youknowone are you suggesting comparing class names as strings character by character? It takes a lot of time on performance.

Comment (Masorubka1) [reply stdlib/src/hashlib.rs:316]: I agree that method that i used is disgusting, but i didn't found any examples by comparing instances enums.

Comment (youknowone) [reply stdlib/src/hashlib.rs:316]: I left a code suggestion using matches! macro

**Review by youknowone — APPROVED** [review]
Thank you! This is great, I like it very much!

**Pull request merged**